apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
-/* global Promise,
-      EventDispatcher
-*/
-
-/* exported Navigation */
+/*global
+  Utils
+ */
+(function(exports) {
 
 'use strict';
 
-(function(window) {
+/**
+ * The Defer type is useful when creating promises.
+ * @typedef {Object} Defer
+ * @property {function(*)} resolve The Promise's resolve function.
+ * @property {function(*)} reject The Promise's reject function.
+ * @property {Promise} promise The actual promise.
+ */
+
+/**
+ * @function
+ * The debug function translates to console.log in debug mode.
+ *
+ * @param {...*} args The values to be displayed to the console.
+ */
+const debug = 1 ? (...args) => console.log(...args) : () => {};
+
+/**
+  * This contains an array of view characteristics, necessary to handle
+  * navigation requests.
+  * Each item has the following properties:
+  * - name: this is the name of the view. This will be used also in the URL
+  *   hash if the same HTML file holds several views.
+  * - url: this is the URL used to match the current location and find the
+  *   current view.
+  * - view: this is the actual JavaScript object that manages this view.
+  * - previous (optional): this is the name of the previous view; this is
```

Comment:
Will it work correctly in case we have different previous panels? Eg.:
* New_Message:
 * conversation ---> new_message  (doesn't exist now);
 * inbox ---> new_message;
 * group ---> new_message (via context menu ---> Send Message).
* Conversation:
 * inbox ---> conversation;
 * new_message ---> conversation (once we sent message, or context menu on number ---> Send Message);
 * goup ---> another conversation (via context menu ---> Send Message).